### ISSUE TITLE
Introduce law checking and property-based testing.

### ISSUE THREAD

**Issue by aloiscochard** [scalaz/scalaz#1230]
We currently lack those.

Scalaprops have been mentioned (https://github.com/scalaprops/scalaprops) and it has the advantage of being maintained by one of `scalaz` contributor :-)

**Comment by jbgi** [comment]
by the way, what about support of both scalaz 7 and 8 in the same classpath? or is osgi the only solution for this situation?
If migrating all dependencies of a project is necessary before being able to try out scalaz8, it will slowdown adoption...

**Comment by vmarquez** [comment]
@aloiscochard @jbgi we could have a different namespace...?

**Comment by xuwei-k** [comment]
I don’t mind forking. Actually I hope embed as a new module (like `"org.scalaz" %% "scalaz-test"` or `"org.scalaz" %% "scalaprops"` ?)

**Comment by aloiscochard** [comment]
@jbgi @vmarquez about namespaces and migrations.

Due to the difference in structure I think it's fine, because in 8 we store thing in `typeclass` and `data` I think we won't run into conflicts even in project using both version (as long as we don't have similar types in `scalaz` but I don't see how that collision could occur if we continue with the current design).

Did I miss something?

@xuwei-k YES! I love the idea and don't have strong opinion about the naming, up to you.

**Comment by jbgi** [comment]
@aloiscochard I think the two package objects, at least, may collide not gracefully:
- https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/package.scala
- https://github.com/scalaz/scalaz/blob/series/8.0.x/base/src/main/scala/package.scala

**Comment by aloiscochard** [comment]
@jbgi good call, thanks!

@jbgi @vmarquez I'll think about it, I'm not so keen on having version number in package name... for example, I would prefer having a specific 7 release that is in a isolated namespace than the other way around.

Maybe that could be done with that seven-backward-compat module we mention a few times... basically a module that gives backward compat on 7 project using 8, I think that would mitigate adoption slowdown and even potentially remove the need of such namespace isolation totally?

**Comment by vmarquez** [comment]
@aloiscochard yeah I'm not too keen on it either, just thought i'd throw it out there as one possible solution that doesn't involve OSGI.  

@xuwei-k yeah that would work, you could embed scala-props in the scalaz8 project, depend on base, then our tests of base could depend on scala-props, is that the idea?
